meilisearch-schema/src/schema.rs

```diff
@@ -0,0 +1,224 @@
+use std::collections::{HashMap, HashSet};
+
+use serde::{Serialize, Deserialize};
+
+use crate::{FieldsMap, FieldId, SResult, Error, IndexedPos};
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Schema {
+    fields_map: FieldsMap,
+
+    identifier: FieldId,
+    ranked: HashSet<FieldId>,
+    displayed: HashSet<FieldId>,
+
+    indexed: Vec<FieldId>,
+    indexed_map: HashMap<FieldId, IndexedPos>,
+
+    must_index_new_fields: bool,
+}
+
+impl Schema {
+
+    pub fn with_identifier<S: Into<String>>(name: S) -> Schema {
+        let mut fields_map = FieldsMap::default();
+        let field_id = fields_map.insert(name.into()).unwrap();
+
+        Schema {
+            fields_map,
+            identifier: field_id,
+            ranked: HashSet::new(),
+            displayed: HashSet::new(),
+            indexed: Vec::new(),
+            indexed_map: HashMap::new(),
+            must_index_new_fields: true,
+        }
+    }
+
+    pub fn identifier(&self) -> String {
+        self.fields_map.get_name(self.identifier).unwrap().to_string()
+    }
+
+    pub fn set_identifier(&mut self, id: String) -> SResult<()> {
+        match self.get_id(id.clone()) {
+            Some(id) => {
+                self.identifier = id;
+                Ok(())
+            },
+            None => Err(Error::FieldNameNotFound(id))
+        }
+    }
+
+    pub fn get_id<S: Into<String>>(&self, name: S) -> Option<FieldId> {
+        self.fields_map.get_id(name)
+    }
+
+    pub fn get_name<I: Into<FieldId>>(&self, id: I) -> Option<String> {
+        self.fields_map.get_name(id)
+    }
+
+    pub fn contains<S: Into<String>>(&self, name: S) -> bool {
+        self.fields_map.get_id(name.into()).is_some()
+    }
+
+    pub fn get_or_create_empty<S: Into<String>>(&mut self, name: S) -> SResult<FieldId> {
+        self.fields_map.insert(name)
+    }
+
+    pub fn get_or_create<S: Into<String> + std::clone::Clone>(&mut self, name: S) -> SResult<FieldId> {
+        match self.fields_map.get_id(name.clone()) {
+            Some(id) => {
+                Ok(id)
+            }
+            None => {
+                if self.must_index_new_fields {
+                    self.set_indexed(name.clone())?;
+                    self.set_displayed(name)
+                } else {
+                    self.fields_map.insert(name.clone())
+                }
+            }
+        }
+    }
```

Comment:
Is it necessary to specify that the type is `Into<String>` and `Clone`, it would be better to convert it in a string and give it to the functions. But are you sure that all those allocations are even needed?